src/runtime/hexagon_remote/qurt_thread_pool.cpp

```diff
@@ -0,0 +1,271 @@
+        // *** Lock work queue ***
+        // ***********************
+        work_queue.lock();
+        if (result) {
+            job->exit_status = result;
+            job->next = job->end;
+            work_queue.jobs = job->next_job;
+        }
+        job->active_workers--;
+        if (job->done()) {
+            if (!owned_job) {
+                qurt_cond_signal(&(job->wakeup_owner));
+            }
+        }
+    }
+    // ***************************
+    // *** Work queue unlocked ***
+    // ***************************
+    work_queue.unlock();
+}
+
+void create_threads(int num_threads) {
+    // Acquire a lock on the work queue.
+    qurt_thread_attr_t thread_attr;
+    for (int i = 0; i < num_threads; ++i) {
+        qurt_thread_attr_init(&thread_attr);
+        qurt_thread_attr_set_stack_addr(&thread_attr, stack[i]);
+        qurt_thread_attr_set_stack_size(&thread_attr, STACK_SIZE);
+        qurt_thread_attr_set_priority(&thread_attr, 100);
+        qurt_thread_create(&threads[i], &thread_attr, (void (*)(void *))worker_thread, NULL);
+    }
+}
+
+void qurt_thread_pool_init() {
+    qurt_sem_init_val(&wait_for_work, 0);
+    work_queue.init();
+    create_threads(NUM_WORKER_THREADS_TO_CREATE);
+    thread_pool_initialized = true;
+}
+
+extern "C" {
+
+int halide_do_par_for(void *user_context, halide_task_t f,
+                      int min, int size, uint8_t *closure) {
+
+    // 1. If the thread pool hasn't been initialized, initiliaze it.
+    // This involves.
+    //    a) Creating the threads.
+    //    b) Acquiring a lock on the work queue and clearing the jobs therein
+    //    c) Setting up a semaphore on which worker threads sleep until awoken
+    //       by this thread i.e. the master thread.
+    if (!thread_pool_initialized) {
+        // Initialize the work queue mutex.
+        // Initialize the wait_for_work semaphore.
+        // lock work queue
+        //    wq.shutdown = false;
+        //    wq.jobs = NULL;
+        // unlcok work queue
+        // create NUM_WORKER_THREADS_TO_CREATE number of threads.
+        // thread_pool_initialized = true;
+        qurt_thread_pool_init();
+    }
+
+    // 2. Lock the work queue again.
+    work_queue.lock();
+
+    // 3. Put work in the global work queue.
+    work job;
+    job.f = f;
+    job.user_context = user_context;
+    job.next = min;          // Start at this index.
+    job.end  = min + size;   // Keep going until one less than this index.
+    job.closure = closure;   // Use this closure.
+    job.exit_status = 0;     // The job hasn't failed yet
+    job.active_workers = 0;  // Nobody is working on this yet
+    qurt_cond_init(&job.wakeup_owner);
+    job.next_job = work_queue.jobs;
+    job.curr_hvx_mode = (qurt_hvx_mode_t) qurt_hvx_get_mode();
+    work_queue.jobs = &job;
+
+    // 4. Unlock global work queue.
+    work_queue.unlock();
+
+    // 5. Wake up the other threads in the pool.
+    qurt_sem_add(&wait_for_work, size);
```

Comment:
This semaphore is just totally buggy. It is incremented by size here, but it only gets decremented when a non-owned job thread runs out of work. What this means is at the end of a job, the worker threads spin to decrement the semaphore down to zero and then sleep. I'm having a hard time seeing a case where it will result in a deadlock, unless it overflows, which is probably possible for large trip count loops. Then down below in the shutdown code, the semaphore is incremented by the number of threads.

I also don't see why it is hard to replace this with a condition variable. I understand a certain degree of just wanting to get the code checked in, but this is both a critical and tricky piece of code. It is prudent to get it right.